### DISCUSSION TITLE
Batch sensor and Stokes integrator.

### DISCUSSION THREAD

**Discussion by frettini** [mitsuba-renderer/mitsuba3#1295] [Q&A]
Hello, 
I am having issues using the batch sensor with the stokes integrator. A problem arise when aligning the stokes vector to the x-axis of the sensor. Instead of aligning to the nested sensor's x-axis, it aligns instead to the parent batch sensor frame. This is probably due to how the sensor is retrieved directly from the scene:
```cpp
auto sensor = scene->sensors()[0];
Vector3f current_basis = mueller::stokes_basis(-ray.d);
Vector3f vertical = sensor->world_transform() * Vector3f(0.f, 1.f, 0.);
```
From what I gathered, this returns the batch sensor's world transform and not the nested sensors' transform. I have tried making world_transform() virtual and make use of m_last_index in the batch sensor to get the nested sensors' transform as such:

```cpp
// batch.cpp
Transform4f world_transform() const override {
        SensorPtr sensor = dr::gather<SensorPtr>(m_sensors_dr, m_last_index);
        return sensor->world_transform();
    }
```
However, this seems to be prone to error in a threaded environment (I have only tested in scalar mode so far). Am I missing something? Do you have any pointers on how I could fix this? Many thanks.

**Comment by njroussel** [comment]
Hi @frettini 

Aha, that's definitely something we hadn't thought about. Does it work if you define `m_last_index` as `thread_local` ? In scalar variants, each thread will be handling one ray at a time. So each thread will need a different `m_last_index` to reflect the individual rays/sensor they have each sampled with `sample_ray`.

**Comment by frettini** [comment]
Hi @njroussel , 
Thanks for the reply and suggestion, it does indeed seem to fix the issue in scalar mode. I was unable to test it in LLVM as I run into the issue raised in #1196 . I do have a similar setup, with radiancemeters nested in a batch sensor.
